### PR TITLE
fix: don't ask for total supply if token doesn't exist

### PR DESCRIPTION
**Pull request by matteojug**
## Description

Closes: https://github.com/stacks-network/sbtc/issues/1026

## Changes

In block observer `update_sbtc_limits`, check token total supply only if we have a total cap and the token contract exists.

Also remove some duplicated code in favor of a dummy const.

## Testing Information

Added an integration test checking that block observer fully process a block in the different scenarios.

## Checklist:

- [ ] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/block_observer.rs [hunk 1]
```diff
@@ -489,18 +490,21 @@ impl<C: Context, B> BlockObserver<C, B> {
...
-                    .context
-                    .get_stacks_client()
-                    .get_sbtc_total_supply(&self.context.config().signer.deployer)
-                    .await?;
-                // The maximum amount of sBTC that can be minted is the total cap
-                // minus the current supply.
-                total_cap.checked_sub(sbtc_supply).unwrap_or(Amount::ZERO)
-            }
+
+        let max_mintable = if limits.total_cap_exists() && self.is_token_deployed().await? {
```
Comment (Jiloc) [inline signer/src/block_observer.rs:494]: Just a thought, but what if we move the `sbtc_contracts_deployed` field from the `TxCoordinatorEventlLoop` to the `SignerState` so that we can query it from here to avoid an extra call to the node?

https://github.com/stacks-network/sbtc/blob/main/signer/src/transaction_coordinator.rs#L1480

**Review by Jiloc — COMMENTED** [review]
PR looks great, I just have a question about a possible change

File: signer/src/testing/stacks.rs [hunk 2]
```diff
@@ -26,6 +28,21 @@ pub const DUMMY_SORTITION_INFO: SortitionInfo = SortitionInfo {
...
+/// Some dummy tenure info
+pub const DUMMY_TENURE_INFO: RPCGetTenureInfo = RPCGetTenureInfo {
+    consensus_hash: ConsensusHash([0; 20]),
+    tenure_start_block_id: StacksBlockId([0; 32]),
+    parent_consensus_hash: ConsensusHash([0; 20]),
+    parent_tenure_start_block_id: StacksBlockId([
+        0x55, 0xc9, 0x86, 0x1b, 0xe5, 0xcf, 0xf9, 0x84, 0xa2, 0x0c, 0xe6, 0xd9, 0x9d, 0x4a, 0xa6,
+        0x59, 0x41, 0x41, 0x28, 0x89, 0xbd, 0xc6, 0x65, 0x09, 0x41, 0x36, 0x42, 0x9b, 0x84, 0xf8,
+        0xc2, 0xee,
+    ]), // StacksBlockId::first_mined()
```
Comment (djordon) [inline signer/src/testing/stacks.rs:40]: Curious, why the change? And what is the significance of this block?

Also, we should remove the comment right?

**Review by djordon — APPROVED** [review]
Looks good ✅

File: signer/src/block_observer.rs (hunk 1 above)
Comment (djordon) [reply signer/src/block_observer.rs:494]: We'd have to also check that the other smart contracts being deployed as well, since this only checks for the token contract. The remaining three smart contracts get submitted after the token one, and they can be confirmed much later.

Comment (djordon) [reply signer/src/block_observer.rs:494]: Oh wait, but this could work if the `tx-coordinator` sets some state variable for each contract when it checks (which is what you had in mind to begin with).

File: signer/src/testing/stacks.rs (hunk 2 above)
Comment (matteojug) [reply signer/src/testing/stacks.rs:40]: We cannot use `StacksBlockId::first_mined()` since is not a const fn; those bytes are the result of calling `StacksBlockId::first_mined()`. I'll nit the comment to avoid confusion about it.

File: signer/src/block_observer.rs (hunk 1 above)
Comment (matteojug) [reply signer/src/block_observer.rs]: Moved to the state -- I've used a single flag, set by the coordinator once all the smart contracts are confirmed as deployed.

File: signer/src/testing/stacks.rs (hunk 2 above)
Comment (matteojug) [reply signer/src/testing/stacks.rs]: Done in [e6ff96b](https://github.com/stacks-network/sbtc/pull/1058/commits/e6ff96b95dc0bc93ff2d531d62eb859e3df1d60c)

**Review by djordon — APPROVED** [review]

**Review by Jiloc — APPROVED** [review]
looks good!
